docs/src/user_docs/assembly/code_organization.md

```diff
@@ -51,12 +42,63 @@ begin
 end
 ```
 
-In addition to local procedures defined within a program, Miden assembly programs can invoke procedures defined in the standard library as described [here](../stdlib/main.md).
+### Modules
+A *module* consists of one or more procedures. There are two types of modules: *library modules* and *executable modules* (also called *programs*).
+
+#### Library modules
+Library modules contain zero or more internal procedures and one or more exported procedures. For example, the following module defines one internal procedure (defined with `proc` instruction) and one exported procedure (defined with `export` instruction):
+```
+proc.foo
+    <instructions>
+end
+
+export.bar
+    <instructions>
+    exec.foo
+    <instructions>
+end
+```
+
+#### Programs
+Executable modules are used to define programs. A program contain zero or more internal procedures (defined with `proc` instruction) and exactly one main procedure (defined with `begin` instruction). For example, the following module defines one internal procedure and a main procedure:
+```
+proc.foo
+    <instructions>
+end
+
+begin
+    <instructions>
+    exec.foo
+    <instructions>
+end
+```
+A program cannot contain any exported procedures.
+
+When a program is executed, the execution starts at the first instruction following the `begin` instruction. The main procedure is expected to be the last procedure in the program and can be followed only by comments.
+
+#### Importing modules
+To invoke a procedure from an external module, the module first needs to be imported using a `use` instruction. Once a module is imported, procedures from this module can be invoked via the regular `exec` or `call` instructions as `exec|call.<module>::<label>` where `label` is the name of the procedure. For example:
 
-### Comments
-Miden assembly allows annotating code with simple comments. Currently, the only supported type of comments is a single-line comment which starts with a `#` (pound) character. For example:
 ```
-# this is a comment
+use.std::math::u64
+
+begin
+    push.1.0
+    push.2.0
+    exec.u64::checked_add
+end
 ```
+In the above example we import `std::math::u64` module from the [standard library](../stdlib/main.md). We then execute a program which pushes two 64-bit integers onto the stack, and then invokes a 64-bit addition procedure from the imported module.
 
-When Miden assembly is serialized into binary format, comments are not retained.
+The set of modules which can be imported by a program can be specified via a Module Provider when instantiating the [Miden Assembler](https://crates.io/crates/miden-assembly) used to compile the program.
+
+### Comments
+Miden assembly allows annotating code with simple comments. There are two types of comments: single-line comment which starts with a `#` (pound) character, and documentation comments which start with `#!` characters. For example:
```

Comment:
nit: We should say "...single-line comments, which start with..." instead of "...single-line comment which starts with..."